azure/compute/src/main/java/org/jclouds/azurecompute/domain/OSImage.java

```diff
@@ -0,0 +1,431 @@
+/**
+ * Copyright (c)Microsoft Open Technologies, Inc.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.azurecompute.domain;
+import com.google.common.base.Objects;
+
+import java.net.URI;
+import java.util.Calendar;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * OS images from the image repository
+ *
+ * @see <a href="http://msdn.microsoft.com/en-us/library/jj157191" >api</a>
+ */
+public class OSImage {
```

Comment:
We use Google Auto for the domain model. This allows us to:

* Enforce immutable objects easily.
* Remove the boilerplate code introduced in large domain objects.
* Avoid adding (and maintaining) the equals/hashcode/tostring methods.

In general, refactor all these domain objects to use Google auto. You can have a look at the latest OpenStack pull requests, ProfitBricks or GCE providers for further reference.